runtimes/pythonrt/dockerclient.go

```diff
@@ -0,0 +1,298 @@
+		if err != nil {
+			return "", err
+		}
+		return n.ID, nil
+	}
+
+	noICCOption := inspectResult.Options["com.docker.network.bridge.enable_icc"]
+	if noICCOption != "false" {
+		return "", errors.New("network with invalid icc, need to recreate")
+	}
+
+	return inspectResult.ID, nil
+}
+
+func (d *dockerClient) StartRunner(runnerImage string, sessionID sdktypes.SessionID, cmd []string) (string, string, error) {
+	resp, err := d.client.ContainerCreate(context.Background(),
+		&container.Config{
+			Image: runnerImage,
+			Cmd:   cmd,
+			Tty:   false,
+
+			ExposedPorts: map[nat.Port]struct{}{
+				nat.Port(internalRunnerPort): {},
+			},
+			Labels: d.runnerLabels,
+		},
+		&container.HostConfig{
+			NetworkMode:  container.NetworkMode(networkName),
+			PortBindings: nat.PortMap{internalRunnerPort: []nat.PortBinding{{HostIP: "0.0.0.0"}}},
+		}, nil, nil, "")
+
+	if err != nil {
+		return "", "", err
+	}
+
+	if err := d.client.ContainerStart(context.Background(), resp.ID, container.StartOptions{}); err != nil {
+		return "", "", err
+	}
+
+	d.mu.Lock()
+	d.allRunnerIDs[resp.ID] = struct{}{}
+	d.mu.Unlock()
+
+	go func() {
+		reader, _ := d.client.ContainerLogs(context.Background(), resp.ID, container.LogsOptions{
+			ShowStdout: true,
+			ShowStderr: true,
+			Follow:     true,
+		})
+		defer reader.Close()
+
+		if d.logRunner {
+			l := d.logger.With(zap.String("runner_id", resp.ID), zap.String("session_id", sessionID.String()))
+			stdourWriter := zapio.Writer{Log: l.With(zap.String("stream", "stdout"))}
+			stderrWriter := zapio.Writer{Log: l.With(zap.String("stream", "stderr"))}
+			_, err = stdcopy.StdCopy(&stdourWriter, &stderrWriter, reader)
+			defer stdourWriter.Close()
+			defer stderrWriter.Close()
+		} else {
+			_, err = io.Copy(io.Discard, reader)
+		}
+
+		if err != nil {
+			d.logger.Warn("error reading runner logs", zap.Error(err), zap.String("runner_id", resp.ID))
+		}
+	}()
+
+	var port string
+
+	for i := 0; i < 10; i++ {
+		inspect, err := d.client.ContainerInspect(context.Background(), resp.ID)
+		if err != nil {
+			return "", "", err
+		}
+
+		ports, ok := inspect.NetworkSettings.Ports[nat.Port(internalRunnerPort)]
+		if ok && len(ports) > 0 {
+			port = ports[0].HostPort
+			break
+		}
+
+		time.Sleep(1000 * time.Millisecond)
+	}
+
+	if port == "" {
+		return "", "", errors.New("couldn't find port")
+	}
+
+	d.mu.Lock()
+	d.activeRunnerIDs[resp.ID] = struct{}{}
+	d.mu.Unlock()
+
+	return resp.ID, port, nil
+}
+
+func (d *dockerClient) SyncCurrentState() error {
+	listedContainers, err := d.client.ContainerList(context.Background(), container.ListOptions{All: true})
+	if err != nil {
+		return err
+	}
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	for _, c := range listedContainers {
+		if _, ok := c.Labels[runnersLabel]; !ok {
+			continue
+		}
+		d.allRunnerIDs[c.ID] = struct{}{}
+
+		if c.State != "running" {
+			continue
```

Comment:
Don't you need to remove it from `activeRunnerIDs` ?

Reply:
done